gop/goputil/goputil.go

```diff
@@ -17,11 +17,147 @@
 package goputil
 
 import (
+	"errors"
+	"go/types"
+	"slices"
+	"sync"
+
 	"github.com/goplus/gop/ast"
 	"github.com/goplus/gop/token"
 	"github.com/goplus/goxlsw/gop"
 )
 
+// InnermostScopeAt returns the innermost scope that contains the given
+// position. It returns nil if not found.
+func InnermostScopeAt(proj *gop.Project, pos token.Pos) *types.Scope {
+	_, typeInfo, _, _ := proj.TypeInfo()
+
+	astFile := PosASTFile(proj, pos)
+	if astFile == nil {
+		return nil
+	}
+
+	fileScope := typeInfo.Scopes[astFile]
+	if fileScope == nil {
+		return nil
+	}
+
+	innermostScope := fileScope
+	for _, scope := range typeInfo.Scopes {
+		if scope.Contains(pos) && fileScope.Contains(scope.Pos()) && innermostScope.Contains(scope.Pos()) {
+			innermostScope = scope
+		}
+	}
+	return innermostScope
+}
+
+// astFileLine represents an AST file line.
+type astFileLine struct {
+	astFile *ast.File
+	line    int
+}
+
+// IdentsAtLine returns the identifiers at the given line in the given AST file.
+func IdentsAtLine(proj *gop.Project, astFile *ast.File, line int) (idents []*ast.Ident) {
+	const cacheKind = "goputil.IdentsAtLine"
```

Comment:
supportedFeat exposes the interface for registering init build to implement registration